numpy/core/src/umath/loops_hyperbolic.dispatch.c.src

```diff
@@ -0,0 +1,395 @@
+        0x3a49f878, 0x39996bf3, 0x388f3e6c, 0x371bb0e3, 0x35a8a5e6, 0x34369b17, 0x322487b0, 0x0,
+        // 7
+        0xbc0e2f66, 0x460bda12, 0x43d638ef, 0xc3e11c3e, 0xc2baa4e9, 0xc249da2d, 0xc1859b82, 0x40dd5b57,
+        0x40494640, 0x40c730a8, 0xbf0f160e, 0x3e30e76f, 0xbea81387, 0xbdb26a1c, 0xbd351e57, 0xbb4c01a0,
+        0x3c1d7bfb, 0x3c722cd1, 0x3c973f1c, 0x3c33a31b, 0x3b862ef4, 0x3a27b3d0, 0xba3b5907, 0xba0efc22,
+        0xb97f9f0f, 0xb8c8af50, 0xb7bdddfb, 0xb64f2950, 0xb4e085b1, 0xb3731dfa, 0xb15a1f04, 0x0
+    };
+
+    const int nlanes = npyv_nlanes_f32;
+    for (; len > 0; len -= nlanes, src += ssrc*nlanes, dst += sdst*nlanes) {
+        npyv_f32 x;
+        if (ssrc == 1) {
+            x = npyv_load_tillz_f32(src, len);
+        } else {
+            x = npyv_loadn_tillz_f32(src, ssrc, len);
+        }
+        npyv_s32 denorm = npyv_and_s32(npyv_reinterpret_s32_f32(x), npyv_setall_s32(0x7fe00000));
+        // |x| > HUGE_THRESHOLD, INF and NaNs are filtered out to callout.
+        npyv_b32 callout_m = npyv_cmple_s32(denorm, npyv_setall_s32(0x7f000000));
+
+        npyv_s32 idxs = npyv_sub_s32(denorm, npyv_setall_s32(0x3d400000));
+                 idxs = npyv_max_s32(idxs, npyv_zero_s32());
+                 idxs = npyv_min_s32(idxs, npyv_setall_s32(0x3e00000));
+        npyv_u32 idx  = npyv_shri_u32(npyv_reinterpret_u32_s32(idxs), 21);
+
+        npyv_f32 b  = npyv_lut32_f32((const float*)lut32x8 + 32*0, idx);
+        npyv_f32 c0 = npyv_lut32_f32((const float*)lut32x8 + 32*1, idx);
+        npyv_f32 c1 = npyv_lut32_f32((const float*)lut32x8 + 32*2, idx);
+        npyv_f32 c2 = npyv_lut32_f32((const float*)lut32x8 + 32*3, idx);
+        npyv_f32 c3 = npyv_lut32_f32((const float*)lut32x8 + 32*4, idx);
+        npyv_f32 c4 = npyv_lut32_f32((const float*)lut32x8 + 32*5, idx);
+        npyv_f32 c5 = npyv_lut32_f32((const float*)lut32x8 + 32*6, idx);
+        npyv_f32 c6 = npyv_lut32_f32((const float*)lut32x8 + 32*7, idx);
+
+        // no need to zerofy nans or avoid FP exceptions by NO_EXC like SVML does
+        // since we're clearing the FP status anyway.
+        npyv_f32 sign = npyv_and_f32(x, npyv_reinterpret_f32_u32(npyv_setall_u32(0x80000000)));
+        npyv_f32 y = npyv_sub_f32(npyv_abs_f32(x), b);
+        npyv_f32 r = npyv_muladd_f32(c6, y, c5);
+        r = npyv_muladd_f32(r, y, c4);
+        r = npyv_muladd_f32(r, y, c3);
+        r = npyv_muladd_f32(r, y, c2);
+        r = npyv_muladd_f32(r, y, c1);
+        r = npyv_muladd_f32(r, y, c0);
+        r = npyv_or_f32(r, sign);
+
+        npy_uint64 callout = npyv_tobits_b32(callout_m);
+        if (NPY_UNLIKELY(callout != ((1 << nlanes) - 1))) {
+            float NPY_DECL_ALIGNED(NPY_SIMD_WIDTH) back2c[npyv_nlanes_f32];
+            npyv_storea_f32(back2c, x);
+            for (unsigned i = 0; i < npyv_nlanes_f32; ++i) {
+                if ((callout >> i) & 1) {
+                    continue;
+                }
+                back2c[i] = npy_tanhf(back2c[i]);
```

Comment:
```suggestion
                // SVML handel |x| > HUGE_THRESHOLD, INF and NaNs as following:
                // 1. check special cases
                // 2. return `+-1` for `|x| > HUGE_THRESHOLD`  otherwise return `x`
```
honestly, I didn't clearly understand why `SVML` did not just handle it through `AVX512` instead of fallback to their scale callout. anyway, I fallback to C since it doesn't worth the risk.

Reply:
+1

Reply:
+1

Reply:
Actually, it looks like the special cases are much easier to handle and perhaps doesn't require a call to the C function. 

1. `np.tanh(x > HUGE_THRESHOLD) = 1.0`
2. `np.tanh(INF) = 1.0`
3. `np.tanh(NAN) = NAN`

It will be more efficient to just write these values to the output register instead of calling C code.

Reply:
@seiko2plus what do you think about using this suggestion?

Reply:
done